File: app/src/main/java/com/salino/sali/data/remote/source/HouseholdRemoteDataSource.kt
package com.salino.sali.data.remote.source

import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.salino.sali.data.model.Household
import com.salino.sali.data.model.HouseholdMember
import com.salino.sali.data.model.MemberRole
import kotlinx.coroutines.tasks.await
import java.util.UUID
import javax.inject.Inject

class HouseholdRemoteDataSource @Inject constructor(
    private val firestore: FirebaseFirestore
) {
    suspend fun createHousehold(userId: String, displayName: String, name: String): Household {
        val householdRef = firestore.collection("households").document()
        val now = Timestamp.now()
        val household = Household(
            id = householdRef.id,
            name = name.trim(),
            createdBy = userId,
            createdAt = now,
            inviteCode = generateInviteCode()
        )
        val member = HouseholdMember(
            userId = userId,
            displayName = displayName,
            role = MemberRole.OWNER.name,
            joinedAt = now
        )
        firestore.runBatch { batch ->
            batch.set(householdRef, household)
            batch.set(householdRef.collection("members").document(userId), member)
            batch.update(firestore.collection("users").document(userId), "activeHouseholdId", household.id)
        }.await()
        return household
    }

    suspend fun joinHousehold(userId: String, displayName: String, inviteCode: String): Household {
        val querySnapshot = firestore.collection("households")
            .whereEqualTo("inviteCode", inviteCode.trim().uppercase())
            .get()
            .await()
        if (querySnapshot.isEmpty) throw IllegalArgumentException("Invalid invite code")

        val householdDoc = querySnapshot.documents.first()
        val household = householdDoc.toObject(Household::class.java)?.copy(id = householdDoc.id)
            ?: throw IllegalStateException("Household not found")
        val member = HouseholdMember(
            userId = userId,
            displayName = displayName,
            role = MemberRole.MEMBER.name,
            joinedAt = Timestamp.now()
        )
        firestore.runBatch { batch ->
            batch.set(householdDoc.reference.collection("members").document(userId), member)
            batch.update(firestore.collection("users").document(userId), "activeHouseholdId", household.id)
        }.await()
        return household
    }

    suspend fun getHousehold(householdId: String): Household {
        val snapshot = firestore.collection("households").document(householdId).get().await()
        return snapshot.toObject(Household::class.java)?.copy(id = snapshot.id)
            ?: throw IllegalStateException("Household not found")
    }

    fun listenToHousehold(householdId: String, onHousehold: (Household) -> Unit): ListenerRegistration {
        return firestore.collection("households").document(householdId)
            .addSnapshotListener { snapshot, _ ->
                val household = snapshot?.toObject(Household::class.java)?.copy(id = snapshot.id) ?: return@addSnapshotListener
                onHousehold(household)
            }
    }

    fun listenToMembers(householdId: String, onMembers: (List<HouseholdMember>) -> Unit): ListenerRegistration {
        return firestore.collection("households").document(householdId).collection("members")
            .addSnapshotListener { snapshot, _ ->
                val members = snapshot?.documents?.mapNotNull { it.toObject(HouseholdMember::class.java) } ?: emptyList()
                onMembers(members)
            }
    }

    private fun generateInviteCode(): String = UUID.randomUUID().toString().take(8).uppercase()

    suspend fun updateHouseholdName(householdId: String, newName: String) {
        firestore.collection("households").document(householdId)
            .update("name", newName.trim())
            .await()
    }

    suspend fun leaveHousehold(householdId: String, userId: String) {
        firestore.runBatch { batch ->
            batch.delete(
                firestore.collection("households").document(householdId)
                    .collection("members").document(userId)
            )
            batch.update(
                firestore.collection("users").document(userId),
                "activeHouseholdId", null
            )
        }.await()
    }
}

File: ios/Remote/HouseholdRemoteDataSource.swift
import Foundation
import FirebaseFirestore

final class HouseholdRemoteDataSource {
    private let firestore: Firestore?

    init(firestore: Firestore? = configuredFirestore()) {
        self.firestore = firestore
    }

    func createHousehold(userId: String, displayName: String, name: String) async throws -> Household {
        guard let firestore else { throw FirebaseConfigurationError.missingGoogleServiceInfo }
        let householdRef = firestore.collection(FirestoreCollections.households).document()
        let now = Date()
        let household = Household(
            id: householdRef.documentID,
            name: name.trimmingCharacters(in: .whitespacesAndNewlines),
            createdBy: userId,
            createdAt: now,
            inviteCode: generateInviteCode()
        )
        let member = HouseholdMember(userId: userId, displayName: displayName, role: .owner, joinedAt: now)
        let batch = firestore.batch()
        batch.setData(household.firestoreData, forDocument: householdRef)
        batch.setData(member.firestoreData, forDocument: householdRef.collection(FirestoreCollections.members).document(userId))
        batch.updateData(["activeHouseholdId": household.id], forDocument: firestore.collection(FirestoreCollections.users).document(userId))
        try await batch.commitAsync()
        return household
    }

    func joinHousehold(userId: String, displayName: String, inviteCode: String) async throws -> Household {
        guard let firestore else { throw FirebaseConfigurationError.missingGoogleServiceInfo }
        let query = firestore.collection(FirestoreCollections.households)
            .whereField("inviteCode", isEqualTo: inviteCode.trimmingCharacters(in: .whitespacesAndNewlines).uppercased())
        let snapshot = try await query.getDocumentsAsync()
        guard let householdDoc = snapshot.documents.first else { throw FirebaseConfigurationError.invalidInviteCode }
        let household = Household(documentId: householdDoc.documentID, data: householdDoc.data())
        let member = HouseholdMember(userId: userId, displayName: displayName, role: .member, joinedAt: Date())
        let batch = firestore.batch()
        batch.setData(member.firestoreData, forDocument: householdDoc.reference.collection(FirestoreCollections.members).document(userId))
        batch.updateData(["activeHouseholdId": household.id], forDocument: firestore.collection(FirestoreCollections.users).document(userId))
        try await batch.commitAsync()
        return household
    }

    func getHousehold(householdId: String) async throws -> Household {
        guard let firestore else { throw FirebaseConfigurationError.missingGoogleServiceInfo }
        let snapshot = try await firestore.collection(FirestoreCollections.households).document(householdId).getDocumentAsync()
        guard snapshot.exists else { throw FirebaseConfigurationError.documentNotFound }
        return Household(documentId: snapshot.documentID, data: snapshot.dataOrEmpty)
    }

    func listenToHousehold(householdId: String, onHousehold: @escaping (Household) -> Void) -> ListenerRegistration {
        guard let firestore else { return NoopListenerRegistration() }
        return firestore.collection(FirestoreCollections.households).document(householdId)
            .addSnapshotListener { snapshot, _ in
                guard let snapshot, snapshot.exists else { return }
                onHousehold(Household(documentId: snapshot.documentID, data: snapshot.dataOrEmpty))
            }
    }

    func listenToMembers(householdId: String, onMembers: @escaping ([HouseholdMember]) -> Void) -> ListenerRegistration {
        guard let firestore else {
            onMembers([])
            return NoopListenerRegistration()
        }
        return firestore.collection(FirestoreCollections.households).document(householdId).collection(FirestoreCollections.members)
            .addSnapshotListener { snapshot, _ in
                let members = snapshot?.documents.map { HouseholdMember(data: $0.data()) } ?? []
                onMembers(members)
            }
    }

    func updateHouseholdName(householdId: String, newName: String) async throws {
        guard let firestore else { throw FirebaseConfigurationError.missingGoogleServiceInfo }
        try await firestore.collection(FirestoreCollections.households)
            .document(householdId)
            .updateDataAsync(["name": newName.trimmingCharacters(in: .whitespacesAndNewlines)])
    }

    func leaveHousehold(householdId: String, userId: String) async throws {
        guard let firestore else { throw FirebaseConfigurationError.missingGoogleServiceInfo }
        let batch = firestore.batch()
        let householdRef = firestore.collection(FirestoreCollections.households).document(householdId)
        batch.deleteDocument(householdRef.collection(FirestoreCollections.members).document(userId))
        batch.updateData(["activeHouseholdId": NSNull()], forDocument: firestore.collection(FirestoreCollections.users).document(userId))
        try await batch.commitAsync()
    }

    private func generateInviteCode() -> String {
        String(UUID().uuidString.replacingOccurrences(of: "-", with: "").prefix(8)).uppercased()
    }
}
